src/Microsoft.TemplateEngine.Edge/Settings/AsyncMutex.cs

```diff
@@ -0,0 +1,93 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable enable
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.TemplateEngine.Edge.Settings
+{
+    /// <summary>
+    /// Helper class to work with <see cref="Mutex"/> in <c>async</c> method, since <c>await</c>
+    /// can switch to different thread and <see cref="Mutex.ReleaseMutex"/> must be called from same thread.
+    /// Hence this helper class.
+    /// </summary>
+    internal sealed class AsyncMutex : IDisposable
+    {
+        private readonly TaskCompletionSource<IDisposable> _taskCompletionSource;
+        private readonly ManualResetEvent _blockReleasingMutex = new ManualResetEvent(false);
+        private readonly string _mutexName;
+        private readonly CancellationToken _token;
+        private readonly Action _unlockCallback;
+        private volatile bool _disposed;
+
+        private AsyncMutex(string mutexName, CancellationToken token, Action unlockCallback)
+        {
+            _mutexName = mutexName;
+            _token = token;
+            _unlockCallback = unlockCallback;
+            _taskCompletionSource = new TaskCompletionSource<IDisposable>();
+            ThreadPool.QueueUserWorkItem(WaitLoop);
+        }
+
+        public static Task<IDisposable> WaitAsync(string mutexName, CancellationToken token, Action unlockCallback)
+        {
+            var mutex = new AsyncMutex(mutexName, token, unlockCallback);
+            return mutex._taskCompletionSource.Task;
+        }
```

Comment:
`WaitAsync` is only here to acquire the mutex, not release it. But we also give the user ability to decide what will happen "when the lock is released". This looks a bit error prone in my opinion. There is also the fact that the callback will be called from the thread owning the mutex, which can be very dangerous if this fact goes unnoticed.

Some not-well-thought, of-the-top-of-my-head suggestions:
I think we should remove this callback from here. Deciding what to do after the mutex is released should be up to the code that is releasing: when calling the dispose. Since we cannot synchronously release the mutex here, the `AsyncMutex` type should implement `IAsyncDisposable` and inside the `DisposeAsync` method we should wait for the thread to release the mutex and then return success.

Reply:
If you look at how callback is used, its used by GlobalSettings, that doesn't have control over when AsyncMutex is disposed, but needs to be notified about releasing of lock, do you have suggestion on how to solve this?

Reply:
One solution may be: instead of keeping a separate `bool _locked` field, we can store the `AsyncMutex` object and check if it was released. `GlobalSettings` would do lock checking like:
`if (_asyncMutex != null && _asyncMutex.IsLocked)`
instead of
`if (this._locked)`
This way, a callback wouldn't be needed.

Reply:
Removed callback and introduced `_mutex` and `IsLocked` as suggested